src/app/api/dashboard/[type]/route.js

```diff
@@ -0,0 +1,194 @@
+      });
+      await updateDoc(doc(db, "users", user.id), {
+        ...element,
+        timestamp: Timestamp.now(),
+        [`roles.${params.type}`]: 0,
+      });
+      SG.send({
+        to: user.email,
+        template_id: process.env.SENDGRID_CONFIRMATION_TEMPLATE,
+        dynamic_template_data: {
+          name: user.name,
+          position: params.type.slice(0, -1).toUpperCase(),
+        },
+      });
+    }
+
+    return res.json({ message: "OK" }, { status: 200 });
+  } catch (err) {
+    return res.json(
+      { message: `Internal Server Error: ${err}` },
+      { status: 500 }
+    );
+  }
+}
+
+export async function GET(req, { params }) {
+  const res = NextResponse;
+  const { auth, message } = await authenticate({
+    admins: [1],
+  });
+
+  if (auth !== 200) {
+    return res.json(
+      { message: `Authentication Error: ${message}` },
+      { status: auth }
+    );
+  }
+
+  const output = [];
+  try {
+    let snapshot;
+    if (types.includes(params.type)) {
+      snapshot = await getDocs(
+        query(
+          collection(db, "users"),
+          where(`roles.${params.type}`, "in", [-1, 0, 1])
+        )
+      );
+      snapshot.forEach((doc) => {
+        const data = doc.data();
+        const element = {};
+        ATTRIBUTES[params.type].forEach((attribute) => {
+          element[attribute] = data[attribute];
+        });
+        output.push({
+          ...element,
+          uid: doc.id,
+          timestamp: data.timestamp,
+          status: data.roles[params.type],
+          selected: false,
+          hidden: false,
+        });
+      });
+    }
+
+    const sorted = output.sort((a, b) =>
+      a.timestamp.seconds < b.timestamp.seconds ? 1 : -1
+    );
+    return res.json({ message: "OK", items: sorted }, { status: 200 });
+  } catch (err) {
+    return res.json(
+      { message: `Internal Server Error: ${err}` },
+      { status: 500 }
+    );
+  }
+}
+
+export async function PUT(req, { params }) {
+  const res = NextResponse;
+  const { objects, status } = await req.json();
+  const { auth, message } = await authenticate(AUTH.PUT);
+
+  if (auth !== 200) {
+    return res.json(
+      { message: `Authentication Error: ${message}` },
+      { status: auth }
+    );
+  }
+  try {
+    if (types.includes(params.type)) {
+      const commits = [];
+      objects.map((object) => {
+        commits.push(
+          updateDoc(doc(db, "users", object.uid), {
+            [`roles.${params.type}`]: status,
+          })
+        );
+        commits.push(
+          SG.send({
+            to: object.email,
+            template_id:
+              status === 1
+                ? process.env.SENDGRID_ACCEPTANCE_TEMPLATE
+                : process.env.SENDGRID_REJECTION_TEMPLATE,
+            dynamic_template_data: {
+              name: object.name,
+              position: params.type.slice(0, -1).toUpperCase(),
+            },
+          })
+        );
+      });
+      await Promise.all(commits);
```

Comment:
keep it regular cause parallel writes is a separate issue, this pr is only concerned with merging API